Fix #2629: Review the encoding of fields in the IR.

sjrd opened the pull request:
Instead of being name-mangled in the IR, their identity now explicitly refers to their declaring class. The mangling is deferred to the emitter.

This increases the code size because the mangling done by the emitter includes the full name of the declaring class, whereas previously we used the number of superclasses in the hierarchy. The new mangling is easier to deal with in incremental runs of the linker. A more efficient stategy could be reintroduced in the future, without breaking compatibility.

The PR includes a few preparatory commits.

sjrd commented:
@gzm0 In the interest of making progress towards 1.0.0, I have simply bumped the figures in checksizes.sh, punting on the code size issue. We'll be able to improve code size in future, compatible releases.

So this is ready for review.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/JSGen.scala:
```diff
@@ -98,11 +98,36 @@ private[emitter] final class JSGen(val semantics: Semantics,
+      implicit pos: Position): Ident = {
+    Ident(cls + "__f_" + field.name, field.originalName)
+  }
+
```
This is all the same than `genSelect`. So I assume this is forward proofing to be able to replace these things with symbols so that they can be hidden from users entirely. However, I wonder if we shouldn't do this type based instead.

sjrd replied:
> So I assume this is forward proofing to be able to replace these things with symbols so that they can be hidden from users entirely.

That's right.

> However, I wonder if we shouldn't do this type based instead.

Hum, I don't understand what you mean. How would we do it type-based? Or *what* exactly would we do type-based?

gzm0 replied:
Seems to pretty much be the same story than above. Do this based on the class kind of `cls`. If it is a Scala class, put a field like now. If it is a JS class, do the fancy symbol (or whatever other thing).

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/checker/IRChecker.scala:
```diff
@@ -941,7 +933,8 @@ private final class IRChecker(unit: LinkingUnit, logger: Logger) {
 
-      case JSPrivateSelect(qualifier, item) =>
+      case JSPrivateSelect(qualifier, cls, field) =>
+        // TODO Should we check that cls::field is defined somewhere?
```
Yes :)

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:
```diff
@@ -4380,6 +4376,45 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
+        this.allFields == that.allFields
+      case _ =>
+        false
+    }
```
Hum, no: `recordFieldNameOf` is a function of `recordFieldNames`, which is itself a function of `recordType`, which is a function of `allFields`. `recordFieldNameOf` deterministically depends on `allFields`, so checking `allFields` is enough.

gzm0 replied:
Oh, I see. `FreshNameAllocator` is entirely deterministic. Fair enough.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:
```diff
@@ -5280,4 +5316,12 @@ private[optimizer] object OptimizerCore {
+  type FieldID = String
+
+  def fieldID(field: Ident): FieldID =
+    field.name
```
Making it an `AnyVal` is going to allocate more and cause more indirections, because it's used quite a lot in type parameters (of `Map`s, notably), so I wouldn't do that. I'd like to make it an `opaque type`, but that doesn't exist yet ;)

I'll make sure to always use appropriate constructors, though.

gzm0 replied:
Well, yes, for two commits. And then you replace it with a tuple anyways. So the `AnyVal` would become a case class and be on par, no?

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala:
```diff
@@ -1792,24 +1773,19 @@ private[emitter] class FunctionEmitter(jsGen: JSGen) {
+                  /* Important: do not emit the assignment, otherwise
+                   * Closure recreates a literal with the duplicate field!
+                   */
+                  value
```
#3670

sjrd commented on ci/checksizes.sh:
```diff
@@ -34,21 +34,21 @@ REVERSI_OPT_GZ_SIZE=$(stat '-c%s' "$REVERSI_OPT.gz")
-    REVERSI_PREOPT_EXPECTEDSIZE=595000
-    REVERSI_OPT_EXPECTEDSIZE=131000
     REVERSI_PREOPT_GZ_EXPECTEDSIZE=70000
     REVERSI_OPT_GZ_EXPECTEDSIZE=30000
```
I am not entirely sure, but I think it could be because field names are more different across unrelated class hierarchies, so there are more distinct names, and that means GCC needs to use more distinct short names, which in turn can cause it to use longer names.

gzm0 replied:
Ah, fair enough. That makes sense.

sjrd commented on ir/src/main/scala/org/scalajs/ir/Trees.scala:
```diff
@@ -488,7 +493,7 @@ object Trees {
   }
 
-  case class JSPrivateSelect(qualifier: Tree, item: Ident)(
+  case class JSPrivateSelect(qualifier: Tree, cls: ClassRef, field: Ident)(
```
The typing rules are significantly different:
* `Select` requires the `qualifier` to be of a subtype of `cls`, whereas `JSPrivateSelect` accepts anything
* The `tpe` of a `Select` is the `ftpe` of the `FieldDef`, but the `tpe` of a `JSPrivateSelect` is always `any`

and the encoding in JS could and will be significantly different as well (to deal with #2680 and #2748).

We don't have the distinction for static methods because the typing rules and the encoding in JS are exactly the same.

gzm0 replied:
> The typing rules are significantly different

OK. But is/should `JSPrivateSelect` ever be used such that the stricter rules are not observed? For example, why should it be OK to `JSPrivateSelect` on anything where the same is not allowed for private methods?

> encoding in JS could and will be significantly different as well

But this is entirely determined by what kind of class the field is defined on, right?

All in all, I just find the non-duality between `FieldDef` and `Select` / `JSPrivateSelect` a bit strange to understand. I'd find a model where `FieldDef`s can be `Select'ed and the encoding of this just happens to be different for JS / non JS classes easier to understand.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/analyzer/Infos.scala:
```diff
@@ -493,6 +493,8 @@ object Infos {
               builder.addInstantiatedClass(cls, ctor.name)
 
+            case Select(_, ClassRef(cls), _) =>
+              builder.addReferencedClass(cls)
```
We should add it if/when we address the
```scala
// TODO Should we check that cls::field is defined somewhere?
```
but if we only use the `cls` for namespacing without checking it in the type system, then there is no need to reference the class.

gzm0 replied:
Ah, ok.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala:
```diff
@@ -577,12 +573,15 @@ private[emitter] class FunctionEmitter(jsGen: JSGen) {
 
-        case Assign(select @ JSPrivateSelect(qualifier, item), rhs) =>
+        case Assign(RecordFieldVarRef(lhs), rhs) =>
+          pushLhsInto(Lhs.Assign(lhs), rhs, tailPosLabels)
```
To follow the definition order of the `Trees` that they match. We consistently preserve that property in the entire codebase.

gzm0 replied:
SG

gzm0 commented:
> It is also allowed for private methods, and for the same reason: there is no way to prevent it.

I'm looking at the IRChecker, and it seems (and this seems correct), that we do not allow private (non-static) methods on a JS class.

So this, to me, explains sufficiently, why there is a difference between methods and fields.

Further, the different semantics indeed would overload the semantics of `Select` too much. So I think you are right, we should keep `JSPrivateSelect`.

sjrd commented:
Rebased, and updated with a new commit that checks that the fields used in `JSPrivateSelect` nodes are declared.

I believe I have now addressed all your comments.

sjrd commented:
Rebased.

gzm0 commented on compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala:
```diff
@@ -721,7 +721,7 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
           implicit val pos = fdef.pos
           val select = fdef.name match {
-            case ident: js.Ident          => js.JSPrivateSelect(selfRef, ident)
+            case ident: js.Ident          => js.JSPrivateSelect(selfRef, encodeClassRef(sym), ident)
```
Seeing the code with the validation makes me wonder if this line here is not a potential problem for binary compatibility: We are basically forcing the linker to emit a private JS field as:

- Something static somewhere (based on the `Ident`).
- Assignments (where the initial assignment is a-priori indistinguishable from any other assignment).

Contrast this to non-local classes where we have the ability to intercept the constructor function (and we do to some extent, especially for ES5 for prototype chaining).

Are we OK with this restriction on how JS private fields are "created"?

sjrd replied:
I don't think this is problematic. Local classes are, by spec, reverse engineered to behave as if the class had never existed in the first place, and the object had been created from the parent class all along. I believe the restriction you're talking about is the semantically correct thing to do.

gzm0 replied:
Hum, but by that definition the parent class cannot be abstract and the implementation cannot access the private field (granted, we could specify `JSPrivate` to mean privately scoped, not private as in class private, which is probably the right thing to do anyways if we want IRCheck to work post opt ever).

Also, this definition does not explain how it is justifiable that the fields end up as static fields on the corresponding abstract JS type (which IIUC, is merely named the same by happenstance and used as a holder for static methods, and not instantiatable).

Writing this, I realize:
- The parent not being abstract is a real requirement so this is WAI.
- If we define JS private fields as "privately scoped" we should maybe put them in the static namespace all the time and merely have them mean: this field needs to be usable. (this might warrant a different tree, but I'm unsure about that).

sjrd replied:
Our private do mean "private to Scala.js" (not visible to JavaScript). Indeed they don't mean private per class.

> the fields end up as static fields on the corresponding abstract JS type

They're actually instance fields (not static) in the abstract JS type. And yes it basically means "this field must be usable".

I'm not 100% happy with the situation, but I don't have a better idea at the moment. Eventually I'd like not to use private fields at all in anonymous classes, and instead generate a *captured* state object to hold the private fields of an anonymous class.

gzm0 replied:
> They're actually instance fields (not static) in the abstract JS type. 

Right, the line `staticMembers += ...` tripped me off (we should probably rename that accumulator).

> Eventually I'd like not to use private fields at all in anonymous classes, and instead generate a captured state object to hold the private fields of an anonymous class.

That sounds reasonable. Do you think this is feasible before 1.0 so we do not have to allow the strange field capability?

Otherwise, we basically have to allow to set any JS private field on any JS value (and it should work). It seems that only allowing to set them on values of the type they refer to is more reasonable (by allowing I mean runtime, not compile time semantics).

Another way to think about it: With the current semantics we couldn't layout a class at definition time (i.e. post capture). That seems not a good thing to commit to (I'm ok with it as an intermediate step).

sjrd replied:
> Do you think this is feasible before 1.0 so we do not have to allow the strange field capability?

I hope so, yes. I would like to be able to remove the instance fields on abstract JS types, and as you say, it would be nice to specify that private JS field selection is undefined behavior if the receiver is not an instance of the class that declares the field. (We could even check that at run-time in fastOpt, actually.)

gzm0 replied:
I have submitted #3777 for this.

sjrd commented:
> Do you think this is feasible before 1.0 so we do not have to allow the strange field capability?

I hope so, yes. I would like to be able to remove the instance fields on abstract JS types, and as you say, it would be nice to specify that private JS field selection is undefined behavior if the receiver is not an instance of the class that declares the field. (We could even check that at run-time in fastOpt, actually.)

gzm0 commented on compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala:
```diff
@@ -660,6 +660,12 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
+           * `JSPrivateSelect` nodes.
+           */
+          if (fdef.name.isInstanceOf[js.Ident])
+            staticMembers += fdef
```
Rename `staticMembers` to `classDefMembers`? Or at the very least comment that the field is *not* static.
